python/orca/tutorial/NCF/pytorch_dataset.py

```diff
@@ -23,69 +23,112 @@
 import scipy.sparse as sp
 
 import torch.utils.data as data
+from sklearn.preprocessing import MinMaxScaler
 
 
 class NCFData(data.Dataset):
-    def __init__(self, features,
+    def __init__(self, data,
                  num_item=0, train_mat=None, num_ng=0):
         super(NCFData, self).__init__()
-        """ Note that the labels are only useful when training, we thus
-            add them in the ng_sample() function.
-        """
-        self.features_ps = features
+        self.data = data
         self.num_item = num_item
         self.train_mat = train_mat
         self.num_ng = num_ng
         self.is_sampling = False
-        self.labels = [1 for _ in range(len(features))]
+        self.data['label'] = [1.0 for _ in range(len(self.data))]
 
     def ng_sample(self):
         self.is_sampling = True
-        self.features_ng = []
-        for x in self.features_ps:
+
+        features_ps = self.data.values[:, :-1].tolist()
+        features_ng = []
+        for x in features_ps:
             u = x[0]
             for t in range(self.num_ng):
                 j = np.random.randint(self.num_item)
                 while (u, j) in self.train_mat:
                     j = np.random.randint(self.num_item)
-                self.features_ng.append([u, j])
-
-        labels_ps = [1 for _ in range(len(self.features_ps))]
-        labels_ng = [0 for _ in range(len(self.features_ng))]
-
-        self.features_fill = self.features_ps + self.features_ng
-        self.labels_fill = labels_ps + labels_ng
+                features_ng.append([u, j])
+        labels_ps = [1.0 for _ in range(len(features_ps))]
+        labels_ng = [0.0 for _ in range(len(features_ng))]
+        features_fill = features_ps + features_ng
+        labels_fill = labels_ps + labels_ng
+        self.data = pd.DataFrame(features_fill,
+                                 columns=["user", "item"], dtype=np.int64)
+        self.data['label'] = labels_fill
+
+    def merge_features(self, users, movies, total_cols):
+        self.data = users.merge(self.data, on='user')
+        self.data = self.data.merge(movies, on='item')
+        self.data = self.data.loc[:, total_cols]
 
     def __len__(self):
-        return (self.num_ng + 1) * len(self.labels)
+        return len(self.data)
 
     def __getitem__(self, idx):
-        features = self.features_fill if self.is_sampling \
-            else self.features_ps
-        labels = self.labels_fill if self.is_sampling \
-            else self.labels
-
-        user = features[idx][0]
-        item = features[idx][1]
-        label = float(labels[idx])
-        return user, item, label
-
-
-def load_dataset(dataset_dir):
-    data_X = pd.read_csv(
+        return tuple(self.data[idx])
+
+
+def load_dataset(dataset_dir, cal_sparse_feats_input_dims=True,
```

Comment:
cal_sparse_feats_input_dims this is not a good argument, better parse to a function